rust/chains/hyperlane-cosmos/src/mailbox/contract.rs

```diff
@@ -0,0 +1,246 @@
+    TxCostEstimate, TxOutcome, H256, U256,
+};
+
+use crate::address::CosmosAddress;
+use crate::grpc::WasmProvider;
+use crate::payloads::general;
+use crate::payloads::mailbox::{
+    GeneralMailboxQuery, ProcessMessageRequest, ProcessMessageRequestInner,
+};
+use crate::types::tx_response_to_outcome;
+use crate::utils::get_block_height_for_lag;
+use crate::{payloads, ConnectionConf, CosmosProvider, Signer};
+
+#[derive(Clone)]
+/// A reference to a Mailbox contract on some Cosmos chain
+pub struct CosmosMailbox {
+    config: ConnectionConf,
+    domain: HyperlaneDomain,
+    address: H256,
+    provider: CosmosProvider,
+}
+
+impl CosmosMailbox {
+    /// Create a reference to a mailbox at a specific Ethereum address on some
+    /// chain
+    pub fn new(
+        conf: ConnectionConf,
+        locator: ContractLocator,
+        signer: Option<Signer>,
+    ) -> ChainResult<Self> {
+        let provider = CosmosProvider::new(
+            locator.domain.clone(),
+            conf.clone(),
+            Some(locator.clone()),
+            signer,
+        )?;
+
+        Ok(Self {
+            config: conf,
+            domain: locator.domain.clone(),
+            address: locator.address,
+            provider,
+        })
+    }
+
+    /// Prefix used in the bech32 address encoding
+    pub fn bech32_prefix(&self) -> String {
+        self.config.get_bech32_prefix()
+    }
+
+    fn contract_address_bytes(&self) -> usize {
+        self.config.get_contract_address_bytes()
+    }
+}
+
+impl HyperlaneContract for CosmosMailbox {
+    fn address(&self) -> H256 {
+        self.address
+    }
+}
+
+impl HyperlaneChain for CosmosMailbox {
+    fn domain(&self) -> &HyperlaneDomain {
+        &self.domain
+    }
+
+    fn provider(&self) -> Box<dyn HyperlaneProvider> {
+        Box::new(self.provider.clone())
+    }
+}
+
+impl Debug for CosmosMailbox {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        todo!()
+    }
+}
+
+#[async_trait]
+impl Mailbox for CosmosMailbox {
+    #[instrument(level = "debug", err, ret, skip(self))]
+    async fn count(&self, lag: Option<NonZeroU64>) -> ChainResult<u32> {
+        let block_height = get_block_height_for_lag(self.provider.grpc(), lag).await?;
+        self.nonce_at_block(block_height).await
+    }
+
+    #[instrument(level = "debug", err, ret, skip(self))]
+    async fn delivered(&self, id: H256) -> ChainResult<bool> {
+        let id = hex::encode(id);
+        let payload = payloads::mailbox::DeliveredRequest {
+            message_delivered: payloads::mailbox::DeliveredRequestInner { id },
+        };
+
+        let delivered = match self
+            .provider
+            .grpc()
+            .wasm_query(GeneralMailboxQuery { mailbox: payload }, None)
+            .await
+        {
+            Ok(v) => {
+                let response: payloads::mailbox::DeliveredResponse = serde_json::from_slice(&v)?;
+
+                response.delivered
+            }
+            Err(err) => {
```

Comment:
as I'm sure you've noticed the cosmos implementation is very ugly, so definitely feel free to make tweaks along the way (as you're already doing 😅). The impl was largely done outside our team and we did our best to clean it up from a functionality POV but less from a code elegance / maintenance POV. But reading this now I can't imagine why we have this match at all instead of just propagating the error

Reply:
Fixed.